traits/trait_set_object.py

```diff
@@ -51,213 +50,518 @@ def __repr__(self):
+        if notifiers is not None:
+            self.notifiers = notifiers
+
+    def notify(self, removed, added):
+        """ Call all notifiers.
+
+        This simply calls all notifiers provided by the class, if any.
+        The notifiers are expected to have the signature::
+
+            notifier(removed, added)
+
+        Any return values are ignored. Any exceptions raised are not
+        handled. Notifiers are therefore expected not to raise any
+        exceptions under normal use.
+
+        Parameters
+        ----------
+        removed : set
+            The items that have been removed.
+        added : set
+            The new items that have been added to the set.
+        """
+        for notifier in self.notifiers:
+            notifier(removed, added)
+
+    # -- set interface -------------------------------------------------------
+
+    def __iand__(self, value):
+        """  Return self &= value.
+
+        Parameters
+        ----------
+        value : set or frozenset
+            A value.
+
+        Returns
+        -------
+        self : TraitSet
+            The updated set.
+        """
+
+        old_set = self.copy()
+        retval = super().__iand__(value)
+        removed = old_set.difference(self)
+
+        if len(removed) > 0:
+            self.notify(removed, set())
+
+        return retval
+
+    def __ior__(self, value):
+        """ Return self |= value.
+
+        Parameters
+        ----------
+        value : set or frozenset
+            A value.
+
+        Returns
+        -------
+        self : TraitSet
+            The updated set.
+        """
+        old_set = self.copy()
+
+        if isinstance(value, (set, frozenset)):
+            value = {self.item_validator(item)
+                     for item in value}
+
+        retval = super().__ior__(value)
+
+        added = self.difference(old_set)
+
+        if len(added) > 0:
+            self.notify(set(), added)
+
+        return retval
+
+    def __isub__(self, value):
+        """ Return self-=value.
+
+        Parameters
+        ----------
+        value : set or frozenset
+            A value.
+
+        Returns
+        -------
+        self : TraitSet
+            The updated set.
+        """
+
+        old_set = self.copy()
+        retval = super().__isub__(value)
+        removed = old_set.difference(self)
+
+        if len(removed) > 0:
+            self.notify(removed, set())
+
+        return retval
+
+    def __ixor__(self, value):
+        """ Return self ^= value.
+
+        Parameters
+        ----------
+        value : set or frozenset
+            A value.
+
+        Returns
+        -------
+        self : TraitSet
+            The updated set.
+        """
+
+        removed = set()
+        added = set()
+        if isinstance(value, (set, frozenset)):
+            values = set(value)
+            removed = self.intersection(values)
+            raw_result = values.difference(removed)
```

Comment:
Nit: `raw_result` is a bit confusing here: I'm not sure what this is the "result" to. How about `raw_added` and `validated_added` instead?